lib/parser.ex

```diff
@@ -0,0 +1,205 @@
+       ) do
+    if Map.has_key?(continued_packets, bitstream_serial_number) do
+      # a special case when there is just one unfinished packet in the page
+      if Enum.empty?(packets) do
+        page_continued_packet =
+          continued_packets[bitstream_serial_number] <> page_continued_packet
+
+        continued_packets =
+          Map.put(continued_packets, bitstream_serial_number, page_continued_packet)
+
+        {packets, continued_packets, page_continued_packet}
+      else
+        [first_packet | rest_of_packets] = packets
+
+        packets = [
+          continued_packets[bitstream_serial_number] <> first_packet | rest_of_packets
+        ]
+
+        continued_packets = Map.delete(continued_packets, bitstream_serial_number)
+        {packets, continued_packets, page_continued_packet}
+      end
+    else
+      {packets, continued_packets, page_continued_packet}
+    end
+  end
+
+  @spec parse_header(data :: binary) ::
+          {:ok, data :: binary, header_type :: integer, bitstream_serial_number :: integer,
+           number_of_page_segments :: integer}
+          | {:error, :need_more_bytes}
+  defp parse_header(
+         <<"OggS", 0, header_type::unsigned-size(8), _granule_position::unsigned-size(64),
+           bitstream_serial_number::little-unsigned-size(32),
+           _page_sequence_number::unsigned-size(32), _crc::unsigned-size(32),
+           number_of_page_segments::unsigned-size(8), rest::binary>>
+       ) do
+    {:ok, rest, header_type, bitstream_serial_number, number_of_page_segments}
+  end
+
+  defp parse_header(<<_header_data::binary-size(27), _rest::binary>>) do
+    raise "Corrupted stream: invalid page header"
+  end
+
+  defp parse_header(_too_short) do
+    {:error, :need_more_bytes}
+  end
+
+  @spec parse_segment_table(data :: binary, page_segments_count :: integer) ::
+          {:ok, data :: binary, list(integer)} | {:error, :need_more_bytes}
+  defp parse_segment_table(data, page_segments_count) do
+    if byte_size(data) < page_segments_count do
+      {:error, :need_more_bytes}
+    else
+      <<segment_table::binary-size(page_segments_count), data::binary>> = data
+      {:ok, data, :binary.bin_to_list(segment_table)}
+    end
+  end
+
+  @spec parse_segments(data :: binary, segment_table :: list(integer)) ::
+          {binary, list(binary), binary | nil}
+  defp parse_segments(data, segment_table) do
+    chunk_fun = fn element, acc ->
+      if element == 255 do
+        {:cont, [element | acc]}
+      else
+        {:cont, Enum.reverse([element | acc]), []}
+      end
+    end
+
+    after_chunk = fn
+      [] -> {:cont, []}
+      acc -> {:cont, acc, []}
+    end
+
+    packets_segments = Enum.chunk_while(segment_table, [], chunk_fun, after_chunk)
+    packets_lengths = Enum.map(packets_segments, &Enum.sum/1)
+
+    {data, packets} = split_packets(data, packets_lengths)
+
+    if(List.last(segment_table) == 255) do
+      {data, Enum.drop(packets, 1), List.last(packets)}
```

Comment:
Shouldn't we drop the last packet? 

```suggestion
      {data, List.delete_at(packets, -1), List.last(packets)}
```